### PR TITLE
[exo-deploy 2] Read service env vars

### PR DESCRIPTION
**Pull request by hugoesthere**
Resolves the production part of https://github.com/Originate/exosphere/issues/429

This PR does two things: read service env vars and pass them in as command line flags instead of writing a `secret.tfvars` file. We avoid writing secrets to disk and having to manage/delete that file this way.

### CODE REVIEW

File: src/types/service_config.go [hunk 1]
```diff
@@ -13,5 +13,31 @@ type ServiceConfig struct {
...
+	Environment     envVars            `yaml:",omitempty"`
 	Production      map[string]string  `yaml:",omitempty"`
 }
+
+type envVars struct {
+	Default     map[string]string `yaml:",omitempty"`
+	Development map[string]string `yaml:",omitempty"`
+	Production  map[string]string `yaml:",omitempty"`
+	Secrets     []string          `yaml:",omitempty"`
+}
```
Comment (charlierudolph) [inline src/types/service_config.go]: Please define this struct in its own file. We are trying to have a single struct per file

File: src/types/service_config.go [hunk 2]
```diff
@@ -13,5 +13,31 @@ type ServiceConfig struct {
...
+	case "production":
+		envVars = s.Environment.Production
+	case "development":
+		envVars = s.Environment.Development
+	}
+	for k, v := range envVars {
+		defaultVars[k] = v
+	}
+	return defaultVars, s.Environment.Secrets
+}
```
Comment (charlierudolph) [inline src/types/service_config.go]: Can we add unit tests for this method?

File: src/terraform/commands.go [hunk 3]
```diff
@@ -17,11 +18,63 @@ func RunInit(deployConfig types.DeployConfig) error {
...
+	envVarsJSON, err := json.Marshal(terraformEnvVars)
+	if err != nil {
+		return "", err
+	}
+	envVarsEscaped, err := json.Marshal(string(envVarsJSON))
+	if err != nil {
+		return "", err
+	}
+	return string(envVarsEscaped), nil
+}
```
Comment (charlierudolph) [inline src/terraform/commands.go]: Can we add unit test these new methods?

**Review by charlierudolph — COMMENTED** [review]
very cool

File: src/types/service_config.go (hunk 1 above)
Comment (hugoesthere) [reply src/types/service_config.go]: What if it's not being used outside of this file? This seems more like a sub-struct of `ServiceConfig`.

File: src/types/service_config.go [hunk 4]
```diff
@@ -13,5 +13,27 @@ type ServiceConfig struct {
...
 }
+
+// GetEnvVars compiles a service's environment variables
+// It overwrites default variables with environemnt specific ones,
+// returning a map of public env vars and a list of private env var keys
+func (s ServiceConfig) GetEnvVars(environment string) (map[string]string, []string) {
+	defaultVars := map[string]string{}
+	for k, v := range s.Environment.Default {
+		defaultVars[k] = v
+	}
```
Comment (charlierudolph) [inline src/types/service_config.go]: Can we create a merge function which takes in two `map[string]string` and merges the second into the first? I think that will help simplify this and even allow us to remove the envVars variable.

File: src/types/service_config.go [hunk 5]
```diff
@@ -13,5 +13,27 @@ type ServiceConfig struct {
...
 	Dependencies    []DependencyConfig `yaml:",omitempty"`
+	Environment     EnvVars            `yaml:",omitempty"`
 	Production      map[string]string  `yaml:",omitempty"`
 }
+
+// GetEnvVars compiles a service's environment variables
+// It overwrites default variables with environemnt specific ones,
+// returning a map of public env vars and a list of private env var keys
+func (s ServiceConfig) GetEnvVars(environment string) (map[string]string, []string) {
+	defaultVars := map[string]string{}
```
Comment (charlierudolph) [inline src/types/service_config.go]: Can we name this result?

File: src/util/map.go [hunk 6]
```diff
@@ -10,3 +10,11 @@ func JoinStringMaps(maps ...map[string]string) map[string]string {
 	}
 	return result
 }
+
+// Merge merges two maps. It overwrites existing keys of map1 if conflicting keys exist
+func Merge(map1 map[string]string, map2 map[string]string) map[string]string {
+	for k, v := range map2 {
+		map1[k] = v
+	}
+	return map1
```
Comment (charlierudolph) [inline src/util/map.go]: I don't think we should return map1. maps are like pointers so map1 gets updated automatically.

File: src/terraform/commands.go [hunk 7]
```diff
@@ -17,11 +18,72 @@ func RunInit(deployConfig types.DeployConfig) error {
...
+	command = append(command, vars...)
+	err = util.RunAndLog(deployConfig.TerraformDir, []string{}, deployConfig.LogChannel, command...)
 	if err != nil {
 		return errors.Wrap(err, "'terraform plan' failed")
 	}
 	return err
 }
+
+// CompileVarFlags compiles the variable flags passed into a Terraform command
+func CompileVarFlags(deployConfig types.DeployConfig, secrets types.Secrets) ([]string, error) {
```
Comment (charlierudolph) [inline src/terraform/commands.go]: Thoughts on moving this and its helper functions into its own file. Then can also put the tests for this in its own file.

**Review by charlierudolph — APPROVED** [review]

**Pull request closed without merging**